File: src/core/contract-ai/types/generation.ts
/**
 * Contract AI - Layer 2: Generation Types (JAK GENEROWAĆ)
 * 
 * Definiuje instrukcje dla LLM jak generować kod.
 * 
 * @version 2.2.0
 * @see todo/14-reclapp-llm-code-generation-spec.md
 */

// ============================================================================
// GENERATION TARGET
// ============================================================================

/**
 * Cele generacji kodu
 */
export type GenerationTarget = 
  | 'api'
  | 'frontend'
  | 'database'
  | 'tests'
  | 'docker'
  | 'all';

/**
 * Priorytet instrukcji
 */
export type InstructionPriority = 'must' | 'should' | 'may';

// ============================================================================
// CODE EXAMPLE
// ============================================================================

/**
 * Przykład kodu do wykorzystania w instrukcjach
 */
export interface CodeExample {
  /** Nazwa przykładu */
  name: string;
  /** Opis przykładu */
  description?: string;
  /** Język programowania */
  language: string;
  /** Kod źródłowy */
  code: string;
}

// ============================================================================
// GENERATION INSTRUCTION
// ============================================================================

/**
 * Instrukcja generacji dla LLM
 * 
 * @example
 * ```typescript
 * const errorHandlingInstruction: GenerationInstruction = {
 *   target: 'api',
 *   priority: 'must',
 *   instruction: 'Implement proper error handling with try-catch and return appropriate HTTP status codes.',
 *   examples: [
 *     {
 *       name: 'Express error handler',
 *       language: 'typescript',
 *       code: `router.get('/', async (req, res) => {
 *         try {
 *           const data = await service.findAll();
 *           res.json(data);
 *         } catch (error) {
 *           res.status(500).json({ error: 'Internal server error' });
 *         }
 *       });`
 *     }
 *   ]
 * };
 * ```
 */
export interface GenerationInstruction {
  /** Cel instrukcji */
  target: GenerationTarget;
  /** Priorytet */
  priority: InstructionPriority;
  /** Treść instrukcji */
  instruction: string;
  /** Przykłady kodu */
  examples?: CodeExample[];
}

// ============================================================================
// PATTERN VARIABLE
// ============================================================================

/**
 * Zmienna do podstawienia w pattern
 */
export interface PatternVariable {
  /** Nazwa zmiennej */
  name: string;
  /** Opis zmiennej */
  description: string;
  /** Typ zmiennej */
  type: 'string' | 'number' | 'boolean' | 'array' | 'object';
  /** Wartość domyślna */
  default?: any;
  /** Czy wymagana */
  required?: boolean;
}

// ============================================================================
// CODE PATTERN
// ============================================================================

/**
 * Wzorzec kodu do naśladowania przez LLM
 * 
 * @example
 * ```typescript
 * const expressRoutePattern: CodePattern = {
 *   name: 'Express Route Pattern',
 *   description: 'Standard pattern for Express.js routes',
 *   appliesTo: ['api'],
 *   template: `
 *     import { Router, Request, Response } from 'express';
 *     
 *     const router = Router();
 *     
 *     router.get('/', async (req: Request, res: Response) => {
 *       try {
 *         const items = await {{service}}.findAll();
 *         res.json(items);
 *       } catch (error) {
 *         res.status(500).json({ error: 'Internal server error' });
 *       }
 *     });
 *     
 *     export default router;
 *   `,
 *   variables: [
 *     { name: 'service', description: 'Service class name', type: 'string', required: true }
 *   ]
 * };
 * ```
 */
export interface CodePattern {
  /** Nazwa wzorca */
  name: string;
  /** Opis wzorca */
  description: string;
  /** Do jakich celów stosować */
  appliesTo: GenerationTarget[];
  /** Szablon kodu */
  template: string;
  /** Zmienne do podstawienia */
  variables?: PatternVariable[];
}

// ============================================================================
// TECHNICAL CONSTRAINT
// ============================================================================

/**
 * Typy ograniczeń technicznych
 */
export type ConstraintType =
  | 'no-external-deps'
  | 'max-file-size'
  | 'naming-convention'
  | 'no-any-types'
  | 'max-function-length'
  | 'max-complexity'
  | 'custom';

/**
 * Ograniczenie techniczne
 * 
 * @example
 * ```typescript
 * const noAnyConstraint: TechnicalConstraint = {
 *   type: 'no-any-types',
 *   rule: 'All code must be properly typed. No "any" types allowed.',
 *   severity: 'error',
 *   autoFix: false
 * };
 * ```
 */
export interface TechnicalConstraint {
  /** Typ ograniczenia */
  type: ConstraintType;
  /** Opis reguły */
  rule: string;
  /** Poziom błędu */
  severity: 'error' | 'warning';
  /** Czy można automatycznie naprawić */
  autoFix?: boolean;
}

// ============================================================================
// TECH STACK
// ============================================================================

/**
 * Konfiguracja backendu
 */
export interface BackendConfig {
  /** Runtime (node, deno, bun) */
  runtime: 'node' | 'deno' | 'bun';
  /** Język */
  language: 'typescript' | 'javascript';
  /** Framework */
  framework: 'express' | 'fastify' | 'koa' | 'hono' | 'none';
  /** Port */
  port: number;
  /** Dodatkowe biblioteki */
  libraries?: string[];
}

/**
 * Konfiguracja frontendu
 */
export interface FrontendConfig {
  /** Framework */
  framework: 'react' | 'vue' | 'svelte' | 'solid' | 'none';
  /** Bundler */
  bundler: 'vite' | 'webpack' | 'esbuild' | 'none';
  /** Stylowanie */
  styling: 'tailwind' | 'css-modules' | 'styled-components' | 'sass' | 'none';
  /** Biblioteka UI */
  uiLibrary?: 'shadcn' | 'mui' | 'chakra' | 'none';
  /** Dodatkowe biblioteki */
  libraries?: string[];
}

/**
 * Konfiguracja bazy danych
 */
export interface DatabaseConfig {
  /** Typ bazy */
  type: 'postgresql' | 'mysql' | 'sqlite' | 'mongodb' | 'in-memory';
  /** ORM */
  orm?: 'prisma' | 'drizzle' | 'typeorm' | 'mongoose' | 'none';
  /** Connection string pattern */
  connectionPattern?: string;
}

/**
 * Stack technologiczny
 * 
 * @example
 * ```typescript
 * const crmTechStack: TechStack = {
 *   backend: {
 *     runtime: 'node',
 *     language: 'typescript',
 *     framework: 'express',
 *     port: 3000
 *   },
 *   frontend: {
 *     framework: 'react',
 *     bundler: 'vite',
 *     styling: 'tailwind',
 *     uiLibrary: 'shadcn'
 *   },
 *   database: {
 *     type: 'in-memory'
 *   }
 * };
 * ```
 */
export interface TechStack {
  /** Konfiguracja backendu */
  backend: BackendConfig;
  /** Konfiguracja frontendu (opcjonalna) */
  frontend?: FrontendConfig;
  /** Konfiguracja bazy danych (opcjonalna) */
  database?: DatabaseConfig;
}

// ============================================================================
// GENERATION LAYER
// ============================================================================

/**
 * Layer 2: Generation - definiuje JAK LLM ma generować kod
 * 
 * @example
 * ```typescript
 * const crmGeneration: GenerationLayer = {
 *   instructions: [
 *     {
 *       target: 'api',
 *       priority: 'must',
 *       instruction: 'Use Express.js with TypeScript. Each entity must have its own route file.'
 *     },
 *     {
 *       target: 'all',
 *       priority: 'must',
 *       instruction: 'All code must be properly typed. No "any" types allowed.'
 *     }
 *   ],
 *   patterns: [
 *     { name: 'Express Route Pattern', ... }
 *   ],
 *   constraints: [
 *     { type: 'no-any-types', rule: '...', severity: 'error' }
 *   ],
 *   techStack: {
 *     backend: { runtime: 'node', language: 'typescript', framework: 'express', port: 3000 }
 *   }
 * };
 * ```
 */
export interface GenerationLayer {
  /** Instrukcje dla LLM */
  instructions: GenerationInstruction[];
  /** Wzorce kodu */
  patterns: CodePattern[];
  /** Ograniczenia techniczne */
  constraints: TechnicalConstraint[];
  /** Stack technologiczny */
  techStack: TechStack;
}

File: reclapp/models/generation.py
"""
Contract AI - Layer 2: Generation Types (HOW)

Defines HOW LLM should generate code - instructions, patterns, constraints.

Mirrors: src/core/contract-ai/types/generation.ts
"""

from enum import Enum
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


# ============================================================================
# INSTRUCTION TYPES
# ============================================================================

class InstructionPriority(str, Enum):
    """Instruction priority levels"""
    MUST = "must"
    SHOULD = "should"
    MAY = "may"
    MUST_NOT = "must-not"


class Instruction(BaseModel):
    """
    Generation instruction for LLM
    
    Example:
        instruction = Instruction(
            target="api",
            priority=InstructionPriority.MUST,
            instruction="Use Express.js with TypeScript"
        )
    """
    target: str
    priority: InstructionPriority
    instruction: str
    rationale: Optional[str] = None


# ============================================================================
# PATTERN TYPES
# ============================================================================

class PatternType(str, Enum):
    """Design pattern types"""
    REPOSITORY = "repository"
    SERVICE = "service"
    CONTROLLER = "controller"
    FACTORY = "factory"
    SINGLETON = "singleton"
    OBSERVER = "observer"
    STRATEGY = "strategy"
    DECORATOR = "decorator"


class Pattern(BaseModel):
    """
    Design pattern specification
    
    Example:
        pattern = Pattern(
            name="Repository Pattern",
            type=PatternType.REPOSITORY,
            description="Data access abstraction layer",
            applies_to=["entity"]
        )
    """
    name: str
    type: PatternType
    description: Optional[str] = None
    appliesTo: Optional[list[str]] = Field(default=None, alias="applies_to")
    template: Optional[str] = None
    
    model_config = {"populate_by_name": True}


# ============================================================================
# CONSTRAINT TYPES
# ============================================================================

class ConstraintType(str, Enum):
    """Constraint types"""
    SECURITY = "security"
    PERFORMANCE = "performance"
    COMPATIBILITY = "compatibility"
    STYLE = "style"
    ARCHITECTURE = "architecture"


class Constraint(BaseModel):
    """
    Generation constraint
    
    Example:
        constraint = Constraint(
            type=ConstraintType.SECURITY,
            name="Input Validation",
            description="All user inputs must be validated",
            enforcement="required"
        )
    """
    type: ConstraintType
    name: str
    description: Optional[str] = None
    enforcement: Literal["required", "recommended", "optional"] = "required"
    check: Optional[str] = None


# ============================================================================
# TECH STACK
# ============================================================================

class BackendTechStack(BaseModel):
    """Backend technology stack configuration"""
    runtime: str = "node"
    language: str = "typescript"
    framework: str = "express"
    port: int = 3000
    orm: Optional[str] = None
    
    
class FrontendTechStack(BaseModel):
    """Frontend technology stack configuration"""
    framework: str = "react"
    language: str = "typescript"
    styling: str = "tailwind"
    port: int = 5173
    

class DatabaseTechStack(BaseModel):
    """Database technology stack configuration"""
    type: Literal["postgresql", "mysql", "sqlite", "mongodb", "in-memory"] = "in-memory"
    host: Optional[str] = None
    port: Optional[int] = None
    name: Optional[str] = None


class TechStack(BaseModel):
    """
    Complete technology stack configuration
    
    Example:
        tech_stack = TechStack(
            backend=BackendTechStack(runtime="node", language="typescript", framework="express", port=3000),
            frontend=FrontendTechStack(framework="react", language="typescript"),
            database=DatabaseTechStack(type="postgresql")
        )
    """
    backend: BackendTechStack = Field(default_factory=BackendTechStack)
    frontend: Optional[FrontendTechStack] = None
    database: DatabaseTechStack = Field(default_factory=DatabaseTechStack)


# ============================================================================
# GENERATION LAYER
# ============================================================================

class GenerationLayer(BaseModel):
    """
    Layer 2: Generation - defines HOW LLM should generate code
    
    Example:
        generation = GenerationLayer(
            instructions=[
                Instruction(target="api", priority=InstructionPriority.MUST, instruction="Use Express.js with TypeScript"),
            ],
            patterns=[
                Pattern(name="Repository", type=PatternType.REPOSITORY, applies_to=["entity"]),
            ],
            constraints=[
                Constraint(type=ConstraintType.SECURITY, name="Input Validation", description="Validate all inputs"),
            ],
            techStack=TechStack(backend=BackendTechStack(port=3000))
        )
    """
    instructions: list[Instruction] = Field(default_factory=list)
    patterns: list[Pattern] = Field(default_factory=list)
    constraints: list[Constraint] = Field(default_factory=list)
    techStack: TechStack = Field(default_factory=TechStack, alias="tech_stack")
    
    model_config = {"populate_by_name": True}
